### PR TITLE
fix(sync): fix intermittent concurrency issues while syncing directories

### PR DESCRIPTION
**Pull request by edvald**
**What this PR does / why we need it**:

If two processes would copy at the same time to a target, one could
delete the files from the other. We now set rsync's temp directory
outside of the target directory, which resolves the problem.

### CODE REVIEW

File: garden-service/src/plugins/kubernetes/hot-reload.ts [hunk 1]
```diff
@@ -296,11 +302,12 @@ export async function syncToService({ ctx, service, hotReloadSpec, namespace, wo
 
     const syncResult = await Bluebird.map(hotReloadSpec.sync, ({ source, target }) => {
       const src = rsyncSourcePath(service.sourceModule.path, source)
-      const destination = `rsync://localhost:${portForward.localPort}/volume/${rsyncTargetPath(target)}`
+      const destination = `rsync://localhost:${portForward.localPort}/volume/root/${rsyncTargetPath(target)}`
+      const tmpDir = `/tmp/${rsyncTargetPath(target)}`.slice(0, -1)
```
Comment (eysi09) [inline garden-service/src/plugins/kubernetes/hot-reload.ts]: Could you add a comment saying that the slice is to remove the trailing slash?

File: garden-service/src/util/string.ts [hunk 2]
```diff
@@ -77,3 +77,10 @@ export function naturalList(list: string[]) {
     return list.slice(0, -1).join(", ") + " and " + list[list.length - 1]
   }
 }
+
+/**
+ * Generate a random string of a specified `length`.
+ */
+export function randomString(length = 8) {
```
Comment (eysi09) [inline garden-service/src/util/string.ts]: This doesn't appear to be used anywhere.

File: garden-service/src/plugins/kubernetes/container/build.ts [hunk 3]
```diff
@@ -140,13 +140,32 @@ const remoteBuild: BuildHandler = async (params) => {
...
   // https://stackoverflow.com/questions/1636889/rsync-how-can-i-configure-it-to-create-target-directory-on-server
   let src = normalizeLocalRsyncPath(`${buildRoot}`) + `/./${module.name}/`
-
   const destination = `rsync://localhost:${syncFwd.localPort}/volume/${ctx.workingCopyId}/`
+  const syncArgs = ["-vrpztgo", "--relative", "--delete", "--temp-dir", "/tmp", src, destination]
 
   log.debug(`Syncing from ${src} to ${destination}`)
 
+  // TODO: remove this after a few releases (from 0.10.15), since this is only necessary for environments initialized
+  //       with 0.10.14 or earlier.
```
Comment (eysi09) [inline garden-service/src/plugins/kubernetes/container/build.ts]: Minor comment but the `with 0.10.14...` part is oddly indented.

File: garden-service/src/util/string.ts (hunk 2 above)
Comment (edvald) [reply garden-service/src/util/string.ts]: No, I thought I needed to generate a temp folder per transfer, but then I didn't. But I figure it's fine to keep this, it's a basic utility we'll probably use.

File: garden-service/src/plugins/kubernetes/container/build.ts (hunk 3 above)
Comment (edvald) [reply garden-service/src/plugins/kubernetes/container/build.ts]: I guess that's a matter of taste :P I'll change it, don't mind either way.

**Comment by edvald** [comment]
Addressed. And yes, tested quickly on Windows (in a VM). Plus this is of course implicitly tested in our CI.

**Pull request merged**